crypto/fipsmodule/sha/sha512.c

```diff
@@ -141,6 +141,63 @@ int SHA512_256_Init(SHA512_CTX *sha) {
   return 1;
 }
 
+OPENSSL_STATIC_ASSERT(SHA512_CHAINING_LENGTH==SHA384_CHAINING_LENGTH,
+                      sha512_and_sha384_have_same_chaining_length)
+OPENSSL_STATIC_ASSERT(SHA512_CHAINING_LENGTH==SHA512_224_CHAINING_LENGTH,
+                      sha512_and_sha512_224_have_same_chaining_length)
+OPENSSL_STATIC_ASSERT(SHA512_CHAINING_LENGTH==SHA512_256_CHAINING_LENGTH,
+                      sha512_and_sha512_256_have_same_chaining_length)
+
+// sha512_init_from_state_impl is the implementation of
+// SHA512_Init_from_state and SHA224_Init_from_state
+// Note that the state h is always SHA512_CHAINING_LENGTH-byte long
+static int sha512_init_from_state_impl(SHA512_CTX *sha, int md_len,
+                                       const uint8_t h[SHA512_CHAINING_LENGTH],
+                                       uint64_t n) {
+  if(n % ((uint64_t) SHA512_CBLOCK * 8) != 0) {
+    // n is not a multiple of the block size in bits, so it fails
+    return 0;
+  }
+
+  OPENSSL_memset(sha, 0, sizeof(SHA512_CTX));
+  sha->md_len = md_len;
+
+  const size_t out_words = SHA512_CHAINING_LENGTH / 8;
+  for (size_t i = 0; i < out_words; i++) {
+    sha->h[i] = CRYPTO_load_u64_be(h);
+    h += 8;
+  }
+
+  sha->Nh = n >> 32;
+  sha->Nl = n & 0xffffffff;
```

Comment:
Oh, I completely missed that `Nl/Nh` are `uint64_t` for SHA-512 as opposed to `uint32_t` everywhere else. Good catch! Since `n` is `uint64_t`, we can just set `Nl` to `n`. I'm adding comments to the `*_get_state` function to indicate that they can only be called if fewer than 2^64 bits (i.e., 2^61 bytes) have been processed.

Reply:
I'll add a unit test to catch this regression on Monday.

Reply:
I've added the unit test.